tests/unit/s2n_record_write_test.c

```diff
@@ -55,18 +55,18 @@ int main(int argc, char *argv[])
 
         /* Test: Alerts include a sane protocol version */
         {
-            DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER),
+            DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT),
```

Comment:
Just wondering why this change is needed?

Reply:
Short version: because I now check for half-close in s2n_negotiate. But on further thought, maybe I shouldn't be checking for half-close in s2n_negotiate.

On line 69, we negotiate the handshake until the SERVER_HELLO message, meaning that the CLIENT_HELLO is send and received. We have only configured an output buffer for this test, so the connection can only send. So if the connection actually tries to read the CLIENT_HELLO, we get an IO error.

Previously, s2n_negotiate wouldn't send OR receive if the connection closed. So after we queued the alert, the call to negotiate would fail with the S2N_ERR_CLOSED error. But now s2n_negotiate checks for half-closed, and writing a close_notify doesn't close the read side, so it would still try to read the CLIENT_HELLO and therefor get an IO error instead of S2N_ERR_CLOSED. To fix this, I just swapped the server out for a client so that it was trying to write instead of trying to read.

I was thinking that this was the desired behavior in case, for example, the server closed its write side immediately after its last flight when the client still had to read that last flight. However, writing that example out for this answer, I'm realizing that wouldn't actually cause any issues, since the client wouldn't read beyond the last handshake message to the close_notify, so wouldn't know that the connection was half-closed. So maybe I don't need to check for half-closed in s2n_negotiate, and I can put this test back how it was 🤔